src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs

```diff
@@ -524,160 +526,164 @@ public async void RetryExecuteAsyncCancel(string cnnString, SqlRetryLogicBasePro
+        #endregion
 
-            retriesCount = 0;
-            Parallel.For(0, concurrentExecution,
-            i =>
+        #region private members
+        private SqlCommand CreateCommand(SqlConnection cnn, SqlRetryLogicBaseProvider provider, int cancelAfterRetries)
+        {
+            cnn.Open();
+            SqlCommand cmd = cnn.CreateCommand();
+            cmd.RetryLogicProvider = provider;
+            cmd.RetryLogicProvider.Retrying += (object s, SqlRetryingEventArgs e) =>
             {
-                using (SqlConnection cnn = new SqlConnection(cnnString))
-                using (SqlCommand cmd = cnn.CreateCommand())
+                Assert.Equal(e.RetryCount, e.Exceptions.Count);
+                Assert.NotEqual(TimeSpan.Zero, e.Delay);
+
+                if (e.RetryCount >= cancelAfterRetries)
                 {
-                    cnn.Open();
-                    cmd.RetryLogicProvider = provider;
-                    cmd.CommandText = query;
-                    Assert.Throws<AggregateException>(() => cmd.ExecuteReader());
+                    e.Cancel = true;
                 }
-            });
-            Assert.Equal(numberOfTries * concurrentExecution, retriesCount + concurrentExecution);
+            };
+            return cmd;
+        }
 
-            if(DataTestUtility.IsNotAzureSynapse())
+        private static void ProcessDataInParallel(string cnnString, SqlRetryLogicBaseProvider provider,
+                                                  string query, int concurrentExecution,
+                                                  Action<SqlCommand> commandAction)
+        {
+            int numberOfTries = provider.RetryLogic.NumberOfTries;
+            int delayCount = 10_000;
+            int retriesCount = 0;
+            var exceptions = new ConcurrentQueue<Exception>();
+
+            ThreadPool.SetMinThreads(concurrentExecution * 2, concurrentExecution * 2);
+            provider.Retrying += (s, e) => Interlocked.Increment(ref retriesCount);             
+            Parallel.For(0, concurrentExecution,
+            i =>
             {
-                retriesCount = 0;
-                Parallel.For(0, concurrentExecution,
-                i =>
+                try
                 {
                     using (SqlConnection cnn = new SqlConnection(cnnString))
                     using (SqlCommand cmd = cnn.CreateCommand())
                     {
                         cnn.Open();
                         cmd.RetryLogicProvider = provider;
-                        cmd.CommandText = query + " FOR XML AUTO";
-                        Assert.Throws<AggregateException>(() => cmd.ExecuteXmlReader());
+                        cmd.CommandText = query;
+
+                        for (int j = 0; j < delayCount; j++)
+                        {
+                            int longQueryOperation = j * j;
+                        }
+
+                        Assert.Throws<Exception>(() => commandAction(cmd));
```

Comment:
`AggregateException` is expected in all the tests. I don't see any reason to use the `Exception` instead of it and adding one level of assertion.

Reply:
This was a problem with Thread Starvation. The first exception being thrown is coming out of the connection (System.Exception), which is well before we attempt our actual testing of the Retry, which is the expected Aggregate.Exception.

Reply:
Just wondering if you have tried without the for-loop delay in the pipeline. I feel this may cause a high-CPU usage. Have you tried Thread.Sleep()?

Reply:
Adjusting the exception handling as mentioned earlier. Will test shortly via Pipeline.